lib/reek/spec/should_reek_of.rb

```diff
@@ -6,34 +6,76 @@ module Spec
     # An rspec matcher that matches when the +actual+ has the specified
     # code smell.
     #
-    class ShouldReekOf        # :nodoc:
-      def initialize(klass, patterns)
-        @klass = klass
-        @patterns = patterns
+    class ShouldReekOf
+      def initialize(smell_category, smell_details_list)
+        @smell_category     = normalize smell_category
+        @smell_details_list = smell_details_list
       end
 
       def matches?(actual)
         @examiner = Examiner.new(actual)
         @all_smells = @examiner.smells
-        @all_smells.any? { |warning| warning.matches?(@klass, @patterns) }
+        smell_details_given? ? match_with_smell_details? : match_without_smell_details?
       end
 
       def failure_message
-        "Expected #{@examiner.description} to reek of #{@klass}, but it didn't"
+        "Expected #{@examiner.description} to reek of #{@smell_category}, but it didn't"
       end
 
       def failure_message_when_negated
-        "Expected #{@examiner.description} not to reek of #{@klass}, but it did"
+        "Expected #{@examiner.description} not to reek of #{@smell_category}, but it did"
+      end
+
+      private
+
+      def smell_details_given?
+        !@smell_details_list.empty?
+      end
+
+      def match_without_smell_details?
+        @all_smells.any? { |warning| warning.matches?(@smell_category) }
+      end
+
+      def match_with_smell_details?
+        @smell_details_list.map do |smell_detail|
+          @all_smells.any? { |warning| warning.matches?(@smell_category, smell_detail) }
```

Comment:
I think we shoult have ReekOf with smell details require that it finds the exact number of smells of that category, just like  SmellOf did. What we should allow is reeking of other smell categories.